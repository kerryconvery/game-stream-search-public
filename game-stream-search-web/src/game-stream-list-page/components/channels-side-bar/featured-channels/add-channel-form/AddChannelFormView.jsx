import React from 'react';
import { string, func, shape, bool } from 'prop-types';
import _trim from 'lodash/trim';
import _get from 'lodash/get';
import _isNil from 'lodash/isNil';
import TextField from '@material-ui/core/TextField';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import FormGroup from '@material-ui/core/FormGroup';
import Button from '@material-ui/core/Button';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormTemplate, { SubmitButton } from '../../../../../shared-components/form/FormTemplate';

const YouTube = 'YouTube';

const AddChannelFormView = ({ formValues, errors, isSaving, onChange, onCancel, onSave }) => {
  const onFormChange = field => (event) => {
    const values = {
      ...formValues,
      [field]: event.target.value,
    };

    onChange(values);
  }

  return (
    <FormTemplate
      title='Add Channel'
      formContent={
        <FormGroup>
          <FormControl margin='dense' error={!_isNil(_get(errors, 'streamPlatform'))}>
            <Select
              data-test-id='stream-platform'
              autoFocus
              value={_get(formValues, 'streamPlatform')}
              onChange={onFormChange('streamPlatform')}
            >
              <MenuItem value='Twitch'>Twitch</MenuItem>
              <MenuItem value='YouTube'>{YouTube}</MenuItem>
              <MenuItem value='DLive'>DLive</MenuItem>
            </Select>
            {!_isNil(_get(errors, 'streamPlatform')) &&
              <FormHelperText>
                {_get(errors, 'streamPlatform')}
              </FormHelperText>
            }
          </FormControl>
          <FormControl margin='dense'>
            <TextField
              id='channel-name'
              label={ _get(formValues, 'streamPlatform', '') === YouTube ? 'Channel id' : 'Channel name'}
              defaultValue={_get(formValues, 'channelName')}
              onChange={onFormChange('channelName')}
              helperText={_get(errors, 'channelName')}
              error={!_isNil(_get(errors, 'channelName'))}
            />
          </FormControl>
        </FormGroup>
      }
      formActions={
        <>
          <SubmitButton submitting={isSaving} onClick={() => onSave(formValues)}>Save</SubmitButton>
          <Button onClick={onCancel}>Cancel</Button>
        </>
      }
    />
  )
}

AddChannelFormView.propTypes = {
  formValues: shape({
    channelName: string,
    streamPlatform: string,
  }),
  errors: shape({
    channelName: string,
  }),
  isSaving: bool.isRequired,
  onChange: func.isRequired,
  onSave: func.isRequired,
  onCancel: func.isRequired,
}

AddChannelFormView.defaultProps = {
  formValues: {
    streamPlatform: 'Twitch',
  },
  errors: {},
}

export const validateForm = ({ streamPlatform, channelName }) => {
  const errors = {};

  if (_trim(channelName) === '')
  {
    errors['channelName'] = streamPlatform === YouTube ?
      'Please enter a channel id' : 'Please enter a channel name';
  }

  return errors;
}

export const mapApiErrorsToFields = (apiErrors) => {
  const fieldErrors = {};

  apiErrors.forEach(error => {
    if (error.errorCode === 'ChannelNotFoundOnPlatform') {
      fieldErrors['channelName'] = error.errorMessage
    }

    if (error.errorCode === 'PlatformServiceIsNotAvailable') {
      fieldErrors['streamPlatform'] = error.errorMessage
    } 
  })

  return fieldErrors;
}

export default AddChannelFormView;