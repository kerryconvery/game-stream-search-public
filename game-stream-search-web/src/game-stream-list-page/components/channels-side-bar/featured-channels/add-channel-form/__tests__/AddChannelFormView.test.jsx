import React from 'react';
import { render, screen } from '@testing-library/react';
import AddChannelFormView, { validateForm } from '../AddChannelFormView';
import '@testing-library/jest-dom/extend-expect';

describe('Add Channel Form', () => {
  it('should match the snapshot', () => {
    const { container } = renderForm(baseProps);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display the channel name field label as channel id when youtube is selected', () => {
    renderForm({ ...baseProps, formValues: { streamPlatform: 'YouTube' } });

    expect(screen.getByText('Channel id')).toBeInTheDocument();
  });

  it('should display the channel name label as channel name when youtube is not selected', () => {
    renderForm({ ...baseProps, formValues: { streamPlatform: 'Twitch' } });

    expect(screen.getByText('Channel name')).toBeInTheDocument();
  });

  it('should display the channel name label as channel name by default', () => {
    renderForm(baseProps);

    expect(screen.getByText('Channel name')).toBeInTheDocument();
  });

  const baseProps = {
    isSaving: false,
    onChange: jest.fn(),
    onSave: jest.fn(),
    onCancel: jest.fn(),
  }

  const renderForm = (props) => {
    return render(<AddChannelFormView {...props} />)
  }
})

describe('Validate add channel form', () => {
  it('should return the channel name field name as channel id when the stream platform is YouTube', () => {
    const errors = validateForm({ streamPlatform: 'YouTube' });

    expect(errors.channelName).toEqual("Please enter a channel id");
  });

  it('should return the channel name field name as channel name when the stream platform is not YouTube', () => {
    const errors = validateForm({ streamPlatform: 'Twitch' });

    expect(errors.channelName).toEqual("Please enter a channel name");
  });
})