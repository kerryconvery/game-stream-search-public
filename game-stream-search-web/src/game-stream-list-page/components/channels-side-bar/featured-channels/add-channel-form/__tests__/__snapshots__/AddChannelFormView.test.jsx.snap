// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Channel Form should match the snapshot 1`] = `
<div
  class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    class="MuiCardHeader-root makeStyles-title-2"
  >
    <div
      class="MuiCardHeader-content"
    >
      <span
        class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Add Channel
      </span>
    </div>
  </div>
  <div
    class="MuiCardContent-root makeStyles-formContent-3"
  >
    <div
      class="MuiFormGroup-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginDense"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
          data-test-id="stream-platform"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            role="button"
            tabindex="0"
          >
            Twitch
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value="Twitch"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginDense"
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="channel-name"
            id="channel-name-label"
          >
            Channel name
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              id="channel-name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiCardActions-root MuiCardActions-spacing"
  >
    <div
      class="makeStyles-root-4"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Save
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
